bs/src/tablecloth.ml

```diff
@@ -6,6 +10,135 @@ let ( << ) (f1 : 'b -> 'c) (f2 : 'a -> 'b) : 'a -> 'c = fun x -> x |> f2 |> f1
+
+  let toList (a: 'a array) : 'a list = Belt.List.fromArray a
+
+  let to_list = toList
+
+  let toIndexedList array = 
+    Belt.Array.reduceReverse array (length array - 1, []) (fun (i, acc) x -> 
+      (i - 1, ((i, x) :: acc)))
+    |> snd
+
+  let to_indexed_list = toIndexedList
+
+  let get ~index array = Belt.Array.get array index 
+
+  let set ~index ~value array = array.(index) <- value
+
+  let sum (a : int array) : int = Belt.Array.reduce a 0 ( + )
+
+  let floatSum (a : float array) : float = Belt.Array.reduce a 0.0 ( +. )
+
+  let float_sum = floatSum
+
+  let filter ~(f : 'a -> bool) (a : 'a array) : 'a array = Belt.Array.keep a f
+
+  let map ~(f : 'a -> 'b) (a : 'a array) : 'b array = Belt.Array.map a f
+
+  let mapWithIndex ~(f : 'int -> 'a -> 'b) (a : 'a array) : 'b array = Belt.Array.mapWithIndex a f
+
+  let map_with_index = mapWithIndex
+
+  let mapi = mapWithIndex
+
+  let map2 ~(f : 'a -> 'b -> 'c) (a : 'a array) (b : 'b array) : 'c array = Belt.Array.zipBy a b f
+
+  let map3 ~(f : 'a -> 'b -> 'c -> 'd) (arrayA : 'a array) (arrayB : 'b array) (arrayC : 'c array) : 'd array =
+    let minLength : int = Belt.Array.reduce [|length arrayB; length arrayC|] (length arrayA) min in
+    Belt.Array.makeBy minLength (fun i -> f arrayA.(i) arrayB.(i) arrayC.(i))
+
+  let flatMap ~(f : 'a -> 'b array) (a : 'a array) : 'b array = Belt.Array.map a f |> Belt.Array.concatMany
+
+  let flat_map = flatMap
+
+  let find ~(f : 'a -> bool) (array : 'a array) : 'a option = 
+    let rec find_loop array ~f ~length i =
+      if i >= length then None
+      else if f array.(i) then Some (array.(i))
+      else find_loop array ~f ~length (i + 1)
+    in
+    find_loop array ~f ~length:(length array) 0 
+
+  let any ~(f : 'a -> bool) (a : 'a array) : bool = Belt.Array.some a f
+
+  let all ~(f : 'a -> bool) (a : 'a array) : bool = Belt.Array.every a f
+
+  let append (a : 'a array) (a' : 'a array)  : 'a array = Belt.Array.concat a a'
+
+  let concatenate (ars : 'a array array) : 'a array = Belt.Array.concatMany ars
+
+  let flatten = concatenate
+
+  let intersperse ~(sep : 'a) (array : 'a array) : 'a array = 
+    Array.init
+      (max 0 (Array.length array * 2 - 1)) 
+      (fun i -> 
+        if i mod 2 <> 0 then sep else array.(i / 2)
+      )
+
+  let slice ~from ?to_ array =
+    let defaultTo = match to_ with 
+      | None -> length array
+      | Some i -> i
+    in
+    let sliceFrom = 
+      if from >= 0 then min (length array) from 
+      else max 0 (min (length array) (length array + from))
+    in    
+    let sliceTo = 
+      if defaultTo >= 0 then min (length array) defaultTo 
+      else max 0 (min (length array) (length array + defaultTo))
+    in    
+    
+    if sliceFrom >= sliceTo then empty else (
+      Array.init (sliceTo - sliceFrom) (fun i -> array.(i + sliceFrom))
+    )
+
+  let foldLeft ~(f : 'a -> 'b -> 'b) ~(initial : 'b) (a : 'a array) : 'b =
+    Belt.Array.reduce a initial (flip f)
+
```

Comment:
flip can have an overhead of more than 300% on lists, i only use it on native/foldr to get all functions uniform.
Think we should stay with the ocaml style of doing things: ~(f : 'b -> 'a -> 'b)

Reply:
Taking a 300% performance hit is a pretty big deal so it looks like we have two options
- Revert to Belt / Base 's type signature
- Re-implement foldLeft / foldRight from scratch to have the Elm signatures.

It's actually kinda a big deal, as making it ergonomic means we have to ensure all of the other modules have API's that permit "non-flipped" usage 

e.g. Having `foldLeft` take `(f : 'a -> 'b -> 'b)` means that we should also make sure that the signature for `Set.add` is `'a -> 'a t -> 'a t`.

This means making a commitment to either:
1. t-last like Elm
2. t-first like Belt
3. t-middle aka all of the optional arguments come first, then the `t` then any non-labelled arguments then the labelled arguments aka the approach of Base / Core.

I would be in favour of either 1 or 3, I think being able to consistently use `|>` is important:
- I am not a huge fan of the mental overhead of having to switch between `->` and `|>` when using Belt
- It actually gets a lot of recognition outside of the OCaml ecosystem thanks to the likes of Elixir and some JS proposals.

Reading that back we should probably turn this into a GitHub issue.

Reply:
Good discussion! Agreed: make an issue and let's get this merged. I'm happy to merge it with the bad-performance order - we can follow up with a grand plan for fixing the performance later.

Reply:
for me the biggest wart of ocaml is the difference the difference between foldl and foldr.
foldl expects a function of type accumulator -> item -> accumulator
foldr expects an function of type item -> accumulator -> accumulator
Belt did the right thing IMHO and unified both to accumulator -> item -> accumulator
we have the chance to do the same and get the same on native if we use flip once on foldr,
which is worse perfomancewise anyway.

Reply:
i like the t-middle approach of Base

Reply:
Ok, I have created #35